### PR TITLE
[DNM] Runechat refactor

### PR DESCRIPTION
**Pull request by S34NW**
# PR requires a TM to ensure the "ghosts can see runechat" bit doesn't murderise the CPU

## What Does This PR Do
Refactors runechat to allow ghosts to see messages in runechat.
~Emotes are now always italicised in runechat.~ No
Ports some TG optimisations and fixes:
https://github.com/tgstation/tgstation/pull/64366
https://github.com/tgstation/tgstation/pull/74388
https://github.com/tgstation/tgstation/pull/69137
https://github.com/tgstation/tgstation/pull/74428

## Why It's Good For The Game
Letting ghosts see runechat would be a solid W.
This potentially has issues with CPU use, but we should really test these changes to get data on it.

## Images of changes
image

## Testing
Spoke as various things, looked.

## Changelog
:cl:
tweak: Tweaked runechat to let ghosts see it
/:cl:

### CODE REVIEW

File: code/controllers/subsystem/runechat.dm [hunk 1]
```diff
@@ -1,234 +1,16 @@
...
-#undef BUCKET_LEN
-#undef BUCKET_POS
-#undef BUCKET_LIMIT
+/datum/controller/subsystem/timer/runechat/fire(resumed)
+	. = ..() //poggers
+	while(message_queue.len)
+		var/datum/callback/queued_message = message_queue[message_queue.len]
+		queued_message.Invoke()
+		message_queue.len--
+		if(MC_TICK_CHECK)
```
Comment (lewcc) [inline code/controllers/subsystem/runechat.dm:15]: Concerned about .length() checks here, or are you leaving them as .len on purpose for possible perf gains?

File: code/datums/chatmessage.dm [hunk 2]
```diff
@@ -68,14 +79,15 @@
 	INVOKE_ASYNC(src, PROC_REF(generate_image), text, target, owner, lifespan, italics, size, symbol)
 
 /datum/chatmessage/Destroy()
+	if(REALTIMEOFDAY < animate_start + animate_lifespan)
+		stack_trace("Del'd before we finished fading, with [(animate_start + animate_lifespan) - REALTIMEOFDAY] time left")
 	if (owned_by)
 		if (owned_by.seen_messages)
```
Comment (Henri215) [inline code/datums/chatmessage.dm:85]:
```suggestion
	if(REALTIMEOFDAY < (animate_start + animate_lifespan))
		stack_trace("Del'd before we finished fading, with [(animate_start + animate_lifespan) - REALTIMEOFDAY] time left")
	if(owned_by)
		if(owned_by.seen_messages)
```

File: code/datums/chatmessage.dm [hunk 3]
```diff
@@ -137,22 +149,54 @@
...
+			// When choosing to update the remaining time we have to be careful not to update the
+			// scheduled time once the EOL has been executed.
+			if (time_spent >= time_before_fade)
+				animate(m.message, pixel_y = m.message.pixel_y + mheight, time = CHAT_MESSAGE_SPAWN_TIME, flags = ANIMATION_PARALLEL)
+				continue
+
+			var/remaining_time = time_before_fade * (CHAT_MESSAGE_EXP_DECAY ** idx++) * (CHAT_MESSAGE_HEIGHT_DECAY ** combined_height)
+			// Ensure we don't accidentially spike alpha up or something silly like that
+			m.message.alpha = m.get_current_alpha(time_spent)
+			if (remaining_time > 0)
```
Comment (Henri215) [inline code/datums/chatmessage.dm:188]:
```suggestion
			if(time_spent >= time_before_fade)
				animate(m.message, pixel_y = m.message.pixel_y + mheight, time = CHAT_MESSAGE_SPAWN_TIME, flags = ANIMATION_PARALLEL)
				continue

			var/remaining_time = time_before_fade * (CHAT_MESSAGE_EXP_DECAY ** idx++) * (CHAT_MESSAGE_HEIGHT_DECAY ** combined_height)
			// Ensure we don't accidentially spike alpha up or something silly like that
			m.message.alpha = m.get_current_alpha(time_spent)
			if(remaining_time > 0)
```

File: code/controllers/subsystem/runechat.dm (hunk 1 above)
Comment (S34NW) [reply code/controllers/subsystem/runechat.dm:15]: len and length aren't quite the same IIRC

**Comment by Contrabang** [comment]
Not a big fan of the italics emotes, makes them harder to read in runechat
image
image

**Comment by DGamerL** [comment]
image
This is how it currently looks in the game, and here's also how it looks normally.
image

**Comment by matttheficus** [comment]
Runechat seems to instantly disappear, only staying above someone's head for a TINY second.

**Comment by S34NW** [comment]
> Not a big fan of the italics emotes, makes them harder to read in runechat image image

Reverting soon

>Runechat seems to instantly disappear, only staying above someone's head for a TINY second.

Logic issue found with timings, will be fixed soon

**Review by SteelSlayer — APPROVED** [review]

**Pull request merged**